src/metabase/api/card.clj

```diff
@@ -170,69 +167,74 @@
-  query ourselves to get the right values, and return a channel where you can listen for the results."
+(defn- query-for-result-metadata [query]
+  ;; for purposes of calculating the actual Fields & types returned by this query we really only need the first
+  ;; row in the results
+  (let [query (-> query
+                  (assoc-in [:constraints :max-results] 1)
+                  (assoc-in [:constraints :max-results-bare-rows] 1)
+                  (assoc-in [:info :executed-by] api/*current-user-id*))]
+    ;; need add the constraints above before calculating hash because those affect the hash
+    ;;
+    ;; (normally middleware takes care of calculating query hashes for 'userland' queries but this is not
+    ;; technically a userland query -- we don't want to save a QueryExecution -- so we need to add `executed-by`
+    ;; and `query-hash` ourselves so the remark gets added)
+    (assoc-in query [:info :query-hash] (qputil/query-hash query))))
+
+(s/defn ^:private validate-or-recalculate-results-metadata
+  "Check that results `metadata` makes sense for `query`. If it doesn't, recalculate and return the correct metadata."
   [query metadata checksum]
-  (let [valid-metadata? (and (results-metadata/valid-checksum? metadata checksum)
-                             (s/validate qr/ResultsMetadata metadata))]
-    (log/info
-     (cond
-       valid-metadata? (trs "Card results metadata passed in to API is VALID. Thanks!")
-       metadata        (trs "Card results metadata passed in to API is INVALID. Running query to fetch correct metadata.")
-       :else           (trs "Card results metadata passed in to API is MISSING. Running query to fetch correct metadata.")))
-    (if valid-metadata?
-      (a/to-chan [metadata])
-      (qp.async/result-metadata-for-query-async query))))
+  (if (results-metadata/valid-checksum? metadata checksum)
+    metadata
+    (binding [qpi/*disable-qp-logging* true]
+      (try
+        ;; try calculating the columns without running the query -- we can do this for MBQL queries
+        (qp/query->expected-cols query)
+        ;; for native queries, we'll need to run the query, but we only need to fetch a single row.
+        (catch Throwable _
+          (let [query* (query-for-result-metadata query)
+                result (qp/process-query query*)]
+            (or (-> result
+                    :data
+                    :results_metadata
+                    :columns)
+                (throw (ex-info (tru "Error running query to determine results metadata")
+                                (if (map? result) result {}))))))))))
 
 (defn check-data-permissions-for-query
-  "Check that we have *data* permissions to run the QUERY in question."
+  "Check that we have *data* permissions to run the `query` in question."
   [query]
   {:pre [(map? query)]}
   (api/check-403 (query-perms/can-run-query? query)))
 
-(defn- save-new-card-async!
+(defn- ^:deprected save-new-card!
```

Comment:
don't have to change it here but it is really nice when deprecated is a version rather than just true. Makes for weird docstrings too under cider-doc:

```
metabase.test.util/round-fingerprint
 [field]
Deprecated in true
```

Reply:
That's not even supposed to be deprecated anymore. Oops. I marked `save-new-card-async!` as deprecated so usages would show up in red in the editor. Then I reworked it to a sync version. I guess I missed removed the tag because it's just "deprected" 😳